rfd/0131-adminitrative-actions-mfa.md

```diff
@@ -0,0 +1,273 @@
+
+### Server configuration
+
+Starting in Teleport 14, MFA for administrative actions will be turned on by
+default for all users with an MFA device registered. This means that when
+a cluster's second factor is set to one of `optional`, `on`, `webauthn`, or
+`otp`, all normal login sessions will require MFA verification to complete
+admin actions.
+
+### Administrative Actions
+
+An administrative action will be primarily defined as an action which modifies a
+resource in a Teleport Cluster directly. Essentially, this covers any action which
+requires an allow rule with a `create`, `update`, or `delete` verb requirement.
+
+For example:
+
+- `tctl rm/create/edit`
+- `tctl users add/rm/reset`
+- Adding a node from the WebUI
+- Modifying a role form the WebUI
+
+#### Usage Event Creation
+
+The `usage_event` resource can be created by any Teleport User via the
+`DefaultImplicitRole`. This will not be considered an admin action.
+
+### Access Request Approval
+
+Submitting an access request approval does not require an `update` allow rule,
+but since it can be used to escalate privileges, it will still be considered an
+admin action.
+
+Submitting an access request denial will remain a non-admin action.
+
+#### User Actions
+
+Some actions which create, update, or delete user specific resources are
+authorized based on the identity of the requesting user rather than their
+allow rules. For example:
+
+- Creating, listing, and deleting a user's own access requests.
+- Adding or removing MFA devices.
+- Changing the user's password.
+
+These actions will continue to be non-administrative actions and will not require
+MFA to complete (if they don't already).
+
+### UX
+
+The UX of this feature will be very similar to Per-session MFA for `tsh`, `tctl`,
+and the WebUI, and Teleport Connect.
+
+When a user performs an admin action, they will be prompted to tap their MFA key.
+Each admin action will require a single unique tap.
+
+```console
+$ tctl rm roles/access
+Tap any security key
+# success
+```
+
+#### Editing resources
+
+With `tctl edit` or the WebUI, it is possible to edit existing resources in a
+text editor. In this case, MFA will be prompted before proceeding to the text
+editor to prevent data loss from a failed MFA challenge.
+
+```console
+$ tctl edit roles/access
+Tap any security key
+# edit role in Vim
+# success
+```
+
+### MFA enforcement
+
+MFA verification will be required to carry out various administrative actions.
+This will be enforced by the Auth Server API for specific endpoints which are
+considered to be [administrative actions](#administrative-actions).
+
+Essentially, this means that an administrative request to the Auth server will
+follow this flow:
+
+1. Create an Auth API client using the user's existing valid certificates.
+1. Use the client to retrieve an MFA challenge for the user with
+   the existing `rpc CreateAuthenticateChallenge`.
+1. Prompt the user to solve the MFA challenge with their MFA device.
```

Comment:
> However, I'm not sure I can see any security benefit in this

It's admittedly a very limited value.  The concern would be somehow an attacker gets access to an unused MFA challenge and uses it for another purpose.  Though granted, in any case I can consider where that's possible it's probably equally easy to just lie to the user and convince them to MFA something they don't realize (just as you mentioned).

That said, unless it's technically prohibitive for any reason I still think we should make this improvement.  Not just for the risk above, but also for potential auditing around how MFA was used.

What do you think?

Reply:
My 2c is that for the "one-shot" challenges we are OK without encoding the purpose. It's a nice follow up, but not a must-have/MVP requirement.

Reply:
@jentfoo Looking back on this, I do agree it's a good idea.

I'm actually running into an unforeseen issue in the implementation with "bulk" admin actions that perform several API requests. For example, adding a users calls `CreateUser` and `CreateResetPasswordToken`. In these cases, we have to prompt for MFA twice one after the other. Ideally, we would only need to prompt for MFA once for a single "action", even if it involves multiple requests.

I'd like to get some initial thoughts on the 2 potential solutions laid out below before updating the RFD.

### Action Scoped MFA verified certs

One solution is to fall back to using MFA verified certs for these edge cases. However, [as stated previously](https://github.com/gravitational/teleport/pull/27588#discussion_r1232494316), this would require us to make the certificates scoped to the specific actions.

I think for an MVP this would not be hard to implement. We could just encode a list of API requests which are allowed into the cert. Then on the server, we could check that the request being authorized is in the list.

### Action Scoped MFA challenge

Another similar solution is to encode the API requests into the MFA challenge request. Then we could pass the same MFA challenge response to each API request, and the server would check that the API request list in the challenge contains the API request being authorized.

However, the underlying change needed here is to let the MFA challenge exist beyond the first validation, instead expiring after a short duration (30s?) to allow subsequent requests of the same admin action to validate MFA. Since the MFA challenge is scoped, I think this may be a viable tradeoff. From a security perspective, I don't think this is any better/worse than MFA verified certificates that allow reuse.

I prefer this solution as it avoids the "two-shot" approach retrieving MFA verified certificates. @codingllama it looks like this is feasible from an implementation perspective, but IYO does it go against MFA principles too strongly, moreso than encoding MFA verification into certs for reuse?

C.C. @rosstimothy

Reply:
I don't think that letting the challenges live a bit longer is a problem, they could already live until timeout for certain scenarios. Also, we can still delete certain kinds of challenges on the spot (ie, "purposeless" or login/authn challenges).

For challenge persistence we have a few other alternatives, like passing a signed/timestamped JWT-like token to the client and using it as "currency", or using an altogether different key to store this info. (I'll have to think on this a bit more to have a stronger opinion.)

For the challenge purpose, I'd rather encode the purpose in the challenge than a list of endpoints, so we have a bit more future flexibility. I also think of the main questions here is how we can tell that the challenge's purpose is fulfilled.

Reply:
> For the challenge purpose, I'd rather encode the purpose in the challenge than a list of endpoints, so we have a bit more future flexibility.

The way I am thinking about this, an admin action from the client's perspective one of:
 - A single admin action endpoint request
 - A string of admin action endpoints requests
 
So too clarify, I meant that the client will provide a list of endpoints that it wants to call. This way, the client has the flexibility to add/remove endpoints that it needs for an admin action.

> I also think of the main questions here is how we can tell that the challenge's purpose is fulfilled.

I think we can avoid this complexity by just making them expire after 30 seconds. An admin action from the client perspective can be limited to a string of MFA-verified requests being performed back to back in quick succession.

If some action has a long pause for some reason, like if we added a `tctl edit roles` which loops through all roles for edits, then I think it is reasonable to require re-verification with MFA. I don't know of any admin actions like this currently.

> For challenge persistence we have a few other alternatives, like passing a signed/timestamped JWT-like token to the client and using it as "currency", or using an altogether different key to store this info. (I'll have to think on this a bit more to have a stronger opinion.)

IMO just letting the client re-pass the MFA challenge response instead of having it exchange it for a token is simpler and avoids the extra round trip. I agree that this could be an interesting/useful route at some point though.

I also don't think we need to be concerned with the mfa challenge stored in `/web/users/<user>/webauthnsessiondata/login` being overwritten, since we can require a client to complete an admin action before continuing to another one. Or since the admin action is defined on the client side, the client probably just needs to combine the admin actions and make an MFA challenge that accommodates both.

Reply:
@Joerger I am happy to hear you're considering this further.  I still think this would be a valuable improvement.

With regards to encoding the endpoints or a "action" / "scope", I feel either are acceptable.  Endpoints seem precise initially, but the actual action entirely depends on the services implementation of the endpoint.  Similar if we encode a "scope" into the cert / mfa request, it is just dependent on the services implementation of what is allowed for that scope.

For that reason I have historically used concise scopes to represent a series of endpoints to keep client side changes minimal and certificates small.

Reply:
I think a "scope" is better than a list of endpoints, so we can decide what that scope means server-side (without client changes and without giving that power to clients).

Reply:
I'm still not sure I fully see the benefit of server-side scopes for client constructed admin actions. If we add scopes, the power still ultimately lies with the client to decide which scope to use.

For example, how would we scope `tctl create -f resources.yaml`, where `resources.yaml` contains several different resource types? Would we just add a `tctl create` scope that covers `update/create` for basically all resources? If we add this, couldn't the client just always use the `tctl create` scope to cover almost any possible admin action?

If we don't add the `tctl create` scope, we would need to request new MFA for every resources in `resources.yaml`, which admittedly may be an acceptable limit.

However, if we want to make scopes server-side anyways, it would make more sense to add new endpoints instead of MFA scopes. For example, instead of adding a `new-user` scope to call `CreateUser` and `CreateResetPasswordToken`, we could just add a `NewUser` endpoint which does both. This moves defining "scope" to the responsibility of the `gRPC` service endpoints instead of a new not-flushed-out scope concept. Client and server side changes are necessary in both cases.

I don't dislike the idea of scopes in general, I just don't think it solves the issue I'm trying to solve. I think just adding a few scopes like "login" and "session-mfa" would be a nice upgrade from our current situation. We could also just add an unconstrained "admin-action" scope for the time being. This would prevent "login" mfa challenges from being used for "admin-actions" and vice versa, and we could continue discarding MFA challenges for the "login" and "session-mfa" scopes. And in the future we could constrain certain endpoints with new scopes.

FYI I'll formalize this into an RFD update this week, thanks for your initial thoughts and any more feedback you have.